src/DynamoCoreWpf/ViewModels/Watch3D/HelixWatch3DViewModel.cs

```diff
@@ -944,24 +951,27 @@ private void DrawGrid()
 
         private void SetGridVisibility()
         {
-            //return if there is nothing to change
-            if (Model3DDictionary.ContainsKey(DefaultGridName) == isGridVisible) return;
-
-            if (isGridVisible)
+            lock (Model3DDictionaryMutex)
             {
-                if (!gridModel3D.IsAttached)
+                //return if there is nothing to change
+                if (Model3DDictionary.ContainsKey(DefaultGridName) == isGridVisible) return;
+
+                if (isGridVisible)
                 {
-                    OnRequestAttachToScene(gridModel3D);
-                }
+                    if (!gridModel3D.IsAttached)
+                    {
+                        OnRequestAttachToScene(gridModel3D);
+                    }
 
-                Model3DDictionary[DefaultGridName] = gridModel3D;
-            }
-            else
-            {
-                Model3DDictionary.Remove(DefaultGridName);
+                    Model3DDictionary[DefaultGridName] = gridModel3D;
+                }
+                else
+                {
+                    Model3DDictionary.Remove(DefaultGridName);
```

Comment:
We do not need to remove the grid from the scene - unless we have to. You can use `gridModel3D.Visibility == Visibility.Hidden`. That will take care of making the object invisible in the scene. Do we have to remove it from the scene?